heat/core/manipulations.py

```diff
@@ -1850,6 +1851,174 @@ def reshape_argsort_counts_displs(
+    >>> ht.roll(a, 1)
+    DNDarray([[19,  0,  1,  2,  3],
+          [ 4,  5,  6,  7,  8],
+          [ 9, 10, 11, 12, 13],
+          [14, 15, 16, 17, 18]], dtype=ht.int32, device=cpu:0, split=None)
+    >>> ht.roll(a, -1, 0)
+    DNDarray([[ 5,  6,  7,  8,  9],
+          [10, 11, 12, 13, 14],
+          [15, 16, 17, 18, 19],
+          [ 0,  1,  2,  3,  4]], dtype=ht.int32, device=cpu:0, split=None)
+    """
+    sanitation.sanitize_in(x)
+
+    if axis is None:
+        return roll(x.flatten(), shift, 0).reshape(x.shape, new_split=x.split)
+
+    # inputs are ints
+    if isinstance(shift, int):
+        if isinstance(axis, int):
+            if x.split is not None and (axis == x.split or (axis + x.ndim) == x.split):
+                # roll along split axis
+                size = x.comm.Get_size()
+                rank = x.comm.Get_rank()
+
+                lshape_map = x.create_lshape_map()[:, x.split]  # local elements along axis
+                cumsum_map = torch.cumsum(lshape_map, dim=0)  # cumulate along axis
+                indices = torch.arange(size)
+                index_map = torch.repeat_interleave(indices, lshape_map)  # index -> process
+
+                # compute index positions
+                index_old = torch.arange(lshape_map[rank])
+                if rank > 0:
+                    index_old += cumsum_map[rank - 1]
+
+                send_index = (index_old + shift) % x.gshape[x.split]
+                recv_index = (index_old - shift) % x.gshape[x.split]
+
+                # exchange arrays
+                recv = torch.empty_like(x.larray)
+                recv_splits = torch.split(recv, 1, dim=x.split)
+                recv_requests = [None for i in range(x.lshape[x.split])]
+
+                for i in range(x.lshape[x.split]):
+                    recv_requests[i] = x.comm.Irecv(
+                        recv_splits[i], index_map[recv_index[i]], index_old[i]
+                    )
+
+                send_splits = torch.split(x.larray, 1, dim=x.split)
+                send_requests = [None for i in range(x.lshape[x.split])]
+
+                for i in range(x.lshape[x.split]):
+                    send_requests[i] = x.comm.Isend(
+                        send_splits[i], index_map[send_index[i]], send_index[i]
+                    )
+
+                for i in range(x.lshape[x.split]):
+                    recv_requests[i].Wait()
+                for i in range(x.lshape[x.split]):
+                    send_requests[i].Wait()
+
+                return DNDarray(recv, x.gshape, x.dtype, x.split, x.device, x.comm, x.balanced)
+
+            # use pytorch if it's not the split axis
+            rolled = torch.roll(x.larray, shift, axis)
+            return DNDarray(
+                rolled,
+                gshape=x.shape,
+                dtype=x.dtype,
+                split=x.split,
+                device=x.device,
+                comm=x.comm,
+                balanced=x.balanced,
+            )
```

Comment:
it would wonderful if these loops could be switched slightly so they are not so nested. That being said, i understand why you did it this way. I am not against keeping this way, its just odd to read in the moment